### ISSUE TITLE
Implement numpy-only functionality natively

### ISSUE THREAD

**Issue by stinodego** [pola-rs/polars#3890]
Once #3861 is merged, there is still some work to be done to make polars run smoothly without numpy. Here's a to do list:

Frame class:
- [x] ~~`__setitem__` -> Use Python sequence constructors~~ **This functionality is deprecated so not worth fixing.**

Series class:
- [x] `__setitem__` -> Handle case where key is list/tuple
- [x] `take` -> Handle list indices
- [ ] `set_at_idx` -> Need to rethink this function, fully relies on numpy
- [x] `pow` / `rpow`
- [x] `sign`
- [x] trigonometry functions (`sin`/`cos`/`tan` & `arcsin`/`arccos`/`arctan`)
- [x] ~~`std`/`var` when ddof is not equal to 1~~  **Will not be supported natively.**

Expr class:
- [x] `sign`
- [x] trigonometry functions (`sin`/`cos`/`tan` & `arcsin`/`arccos`/`arctan`)

Misc:
- [x] In `Frame`, separate the numpy functionality from `_from_records` into a separate `_from_numpy` method.

**Comment by ritchie46** [comment]
This is much better than I anticipated. I thought we still used numpy for our `Series` constructors, but that is not the case.

Some thoughts:

* For `__setitem__` we can convert the `list` and `tuple` to a `Series` and recursively call `__setitem__` again (or call the proper dispatch function).

* For `take` we can convert a list to a `Series` and that to a `literal` expression and use the `if isinstance(indices, pli.Expr):` branch.

* For `pow`/`powr` we can dispatch via expressions similar to what we did with `std` and `var`.

* `ddof != 1`: I don't think we should support this for now (without numpy that is). As al our variance derived functions assume that `ddof` and so it is consistently not supported. :P

**Comment by stinodego** [comment]
I'll give this a try somewhere in the upcoming week!

**Comment by stinodego** [comment]
I've looked at the `set_at_idx` stuff, but to be honest, the connection between Rust and Python is still quite a mystery to me. Maybe we can look at this together sometime? I would be interested to learn more about the inner workings here. I could swing by the Xomnia office or something :)

Regarding the trigonometry stuff: as far as I can see, Rust polars does not implement any of these functions, correct? So if there is no native implementation here, I think it's fine to rely on numpy for now.

**Comment by ritchie46** [comment]
> I've looked at the set_at_idx stuff, but to be honest, the connection between Rust and Python is still quite a mystery to me. Maybe we can look at this together sometime? I would be interested to learn more about the inner workings here. I could swing by the Xomnia office or something :)

Yes, that's a good idea. It is not the easiest to get started with. But I think it's a nice full introduction to arrow, polars and the crate setup all the way to dispatch in python.

> Regarding sign and the trigonometry functions: as far as I can see, Rust polars does not implement any of these functions, correct? So if there is no native implementation here, I think it's fine to rely on numpy for now. We can make a separate issue for implementing these.

That's correct. In rust it is a bit wasteful to compile these if you don't need them. And because users code in rust, they can simply write a closure and have fast code.

This is something we either only need to add on the python side, or only need to implement in lazy under feature flags. 
I could do one and then that can be cookiecutter for the others?

**Comment by stinodego** [comment]
> This is something we either only need to add on the python side, or only need to implement in lazy under feature flags. I could do one and then that can be cookiecutter for the others?

If we want to keep it on the Python side only, I think we can just keep the NumPy implementation. It's probably faster than whatever we come with ourselves. But if we can dispatch to a fast Rust implementation, that would be preferable.

If you want to implement one, I don't mind doing the rest!

**Comment by ritchie46** [comment]
>  It's probably faster than whatever we come with ourselves

No, it isn't. We match or supersede numpy on most element-wise operations we have written. I will come up with an example.

**Comment by stinodego** [comment]
I briefly looked at the `sign` method, but unfortunately the numpy implementation doesn't seem to have a direct counterpart in Rust.

- [numpy.sign](https://numpy.org/doc/stable/reference/generated/numpy.sign.html)
- [num::Float::signum](https://docs.rs/num/0.1.41/num/trait.Float.html#tymethod.signum)

The main difference is that numpy has a special case for `0` and `-0` (results in 0), where the Rust version just returns 1.0 and -1.0, respectively. Also, the numpy version has support for complex numbers, it seems.

Do you have a suggestion on how to solve this? On option would be to simply switch to the Rust version, but that would be a breaking change.

**Comment by stinodego** [comment]
Closing this issue in favor of #4197.
